### ISSUE TITLE
Rails42: Serialized columns are not properly serialized

### ISSUE THREAD

**Issue by apeiros** [rsim/oracle-enhanced#549]
Serialized columns are serialized using value.to_s instead of the proper serializer:
```
class Foo < ActiveRecord::Base
  serialize :bar
end
created = Foo.create bar: {"some" => "value"}
Foo.find(created.id).bar # => "{\"some\"=>\"value\"}"
```

I suspect this commit is to blame: https://github.com/rsim/oracle-enhanced/commit/6e3ee643469f3fcc21b02cec87e0c9532b98f679
The value is obviously not a string, and due to the condition, it will not be converted to yaml.

Additionally, using to_yaml hard coded is probably wrong too, since rails 4.2 allows custom serializers.

**Comment by apeiros** [comment]
I suggest to change https://github.com/rsim/oracle-enhanced/blob/rails42/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L804 from:
```value = value.to_yaml if value.is_a?(String) && klass.serialized_attributes[col.name]```
to:
```value = col.cast_type.type_cast_for_database(value)```

This should work for other serializers too. However, I have not yet figured out how to properly adapt the reading part. Any input on that would be appreciated.

**Comment by yahonda** [comment]
Thanks for the investigation.
Would you open a pull request for your change? Then we can test easily your commit.

**Comment by apeiros** [comment]
Will do. Currently working on setting up the test environment. Got to solve the "corporate obstacle course" :-S

**Comment by apeiros** [comment]
OK, I'm at a loss. When I run the tests I get "392 examples, 249 failures ". I assume this is not normal.

Here is what I did:
* I set up an oracle 12, largely following https://dimitrisli.wordpress.com/2012/08/08/how-to-install-oracle-database-on-mac-os-any-version/
* The VirtualBox image is from http://www.oracle.com/technetwork/database/enterprise-edition/databaseappdev-vm-161299.html
* I created the oracle_enhanced user according to RUNNING_TESTS.md
* I changed the Gemfile so it uses git over https:

        source 'http://rubygems.org'

        group :development do
          gem 'jeweler', '~> 1.8'
          gem 'rspec', '~> 2.4'
          gem 'rdoc'

          gem 'activerecord',   git: 'https://github.com/rails/rails.git'
          gem 'activemodel',    git: 'https://github.com/rails/rails.git'
          gem 'activesupport',  git: 'https://github.com/rails/rails.git'
          gem 'actionpack',     git: 'https://github.com/rails/rails.git'
          gem 'railties',       git: 'https://github.com/rails/rails.git'

          gem 'arel',           git: 'https://github.com/rails/arel.git'
          gem 'journey',        git: 'https://github.com/rails/journey.git'

          gem 'activerecord-deprecated_finders'
          gem 'ruby-plsql', '>=0.5.0'

          platforms :ruby do
            gem 'ruby-oci8',    git: 'https://github.com/kubo/ruby-oci8.git'
          end
        end
* I changed spec_helper to set host/port explicitly and use the sys & system password of the oracle vm image

        DATABASE_NAME = ENV['DATABASE_NAME'] || 'orcl'
        DATABASE_HOST = "<...vm ip...>" #ENV['DATABASE_HOST']
        DATABASE_PORT = 1521 #ENV['DATABASE_PORT']
        DATABASE_USER = ENV['DATABASE_USER'] || 'oracle_enhanced'
        DATABASE_PASSWORD = ENV['DATABASE_PASSWORD'] || 'oracle_enhanced'
        DATABASE_SYS_PASSWORD = ENV['DATABASE_SYS_PASSWORD'] || 'oracle'
* I verified that I can properly connect with the above settings
* I executed the following statements:

        bundle install
        export RAILS_GEM_VERSION=4.0.13
        rake spec

Is there anything I did wrong?

**Comment by apeiros** [comment]
Here is the full output: https://gist.githubusercontent.com/apeiros/3f6aa3ff5b3570b4cc37/raw/ora_spec_output

**Comment by apeiros** [comment]
So apparently the change to the Gemfile prevents the version selection from working. Inspecting the errors I noticed AR version was 5.0.0-alpha. I changed the Gemfile to use rails 4.0.13 specifically (`gem 'activerecord',   '=4.1.8'` etc.). But I still get "392 examples, 70 failures", which seems to be an excessively large amount of failures. Note: this is against your master branch. No changes on my part other than the above mentioned changes on spec_helper and Gemfile.
Any pointers are appreciated.

**Comment by yahonda** [comment]
Thanks for testing

Oracle enhanced `rails42` branch should work with Rails `v4.2.0` tag or `4-2-stable` branch.

Then there is a repository for Oracle enhanced adapter testing https://github.com/yahonda/rails-dev-box-runs-oracle

Here are step by step commands to execute Oracle enhanced adapter `rails42` branch testing with Rails `v4.2.0`.

https://github.com/yahonda/rails-dev-box-runs-oracle#execute-oracle-enhanced-adapter-unit-tests-with-rails-42

hope this helps.

**Comment by apeiros** [comment]
Ah, that repository would have helped. I could determine that most of the problems were due to me using an oracle 12c server. I set up a virtual box with 11g and the number of errors (master with rails 4.1) went down to less than 10, all related to manual create statements.
With rails42 branch + my commits and rails 4.2 there were a couple more, I think around 20 (sorry for being vague, I'm not at work anymore)

**Comment by jestraw** [comment]
By the way, that original code is also responsible for issue #315 -- that write_lobs method hard-codes serialization to YAML rather than obeying the selected coder.

**Comment by yahonda** [comment]
I think it is fixed by #657. Please reopen if it still reproduce using 1.6.2.

**Comment by tzumby** [comment]
Hey guys,

This is still causing an issue with Oracle 12c, Rails 4.2.1 on Ruby 2.2.0. 

It does work when I change I tried changing this line:

value = value.to_yaml if klass.serialized_attributes[col.name]

to this (what @apeiros originally suggested):

value = col.cast_type.type_cast_for_database(value)

I will post the steps to reproduce this but while I get that going, was there a reason why that initial fix didn't make it to master ?

Cheers

**Comment by yahonda** [comment]
@tzumby Thanks for testing. Merged #688 to master. Please have a look at this if you are still getting issues please open a new one.
